diff -urN tlslite-ng.ecdsa/tlslite/tlsconnection.py tlslite-ng/tlslite/tlsconnection.py
--- tlslite-ng.ecdsa/tlslite/tlsconnection.py	2019-09-27 09:23:03.270984853 -0500
+++ tlslite-ng/tlslite/tlsconnection.py	2019-09-27 09:46:25.294519226 -0500
@@ -695,7 +695,9 @@
                 if cipher in CipherSuite.ecdhAllSuites), None) is not None:
             groups.extend(self._curveNamesToList(settings))
             extensions.append(ECPointFormatsExtension().\
-                              create([ECPointFormat.uncompressed]))
+                              create([ECPointFormat.uncompressed,
+                                      ECPointFormat.ansiX962_compressed_prime,
+                                      ECPointFormat.ansiX962_compressed_char2]))
         # Advertise FFDHE groups if we have DHE ciphers
         if next((cipher for cipher in cipherSuites
                  if cipher in CipherSuite.dhAllSuites), None) is not None:
@@ -2154,7 +2156,9 @@
             # even though the selected cipher may not use ECC, client may want
             # to send a CA certificate with ECDSA...
             extensions.append(ECPointFormatsExtension().create(
-                [ECPointFormat.uncompressed]))
+                [ECPointFormat.uncompressed,
+                 ECPointFormat.ansiX962_compressed_prime,
+                 ECPointFormat.ansiX962_compressed_char2]))
 
         # if client sent Heartbeat extension
         if clientHello.getExtension(ExtensionType.heartbeat):
