--- front/node_modules/jquery-comments_brainkit/js/jquery-comments.js	2017-06-23 18:24:51.000000000 -0500
+++ front/node_modules/jquery-comments_brainkit/js/jquery-comments.js	2020-05-26 14:23:42.344564325 -0500
@@ -2232,11 +2232,11 @@
 
             // URLs starting with http://, https://, file:// or ftp://
             replacePattern1 = /(^|\s)((https?|ftp|file):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gim;
-            replacedText = inputText.replace(replacePattern1, '$1<a href="$2" target="_blank">$2</a>');
+            replacedText = inputText.replace(replacePattern1, '$1<a href="$2" target="_blank" rel="noreferrer noopener">$2</a>');
 
             // URLs starting with "www." (without // before it, or it'd re-link the ones done above).
             replacePattern2 = /(^|\s)(www\.[\S]+(\b|$))/gim;
-            replacedText = replacedText.replace(replacePattern2, '$1<a href="http://$2" target="_blank">$2</a>');
+            replacedText = replacedText.replace(replacePattern2, '$1<a href="http://$2" target="_blank" rel="noreferrer noopener">$2</a>');
 
             // Change email addresses to mailto:: links.
             replacePattern3 = /(^|\s)(([a-zA-Z0-9\-\_\.]+)@[a-zA-Z\_]+?(\.[a-zA-Z]{2,6})+)/gim;
@@ -2251,7 +2251,7 @@
                 var splitInput = inputText.split(/(<\/a>)/g);
                 for (var i = 0; i < splitInput.length; i++) {
                     if (splitInput[i].match(/<a href/g) == null) {
-                        splitInput[i] = splitInput[i].replace(replacePattern1, '<a href="$1" target="_blank">$1</a>').replace(replacePattern2, '$1<a href="http://$2" target="_blank">$2</a>').replace(replacePattern3, '<a href="mailto:$1">$1</a>');
+                        splitInput[i] = splitInput[i].replace(replacePattern1, '<a href="$1" target="_blank" rel="noreferrer noopener">$1</a>').replace(replacePattern2, '$1<a href="http://$2" target="_blank" rel="noreferrer noopener">$2</a>').replace(replacePattern3, '<a href="mailto:$1">$1</a>');
                     }
                 }
                 var combinedReplacedText = splitInput.join('');
